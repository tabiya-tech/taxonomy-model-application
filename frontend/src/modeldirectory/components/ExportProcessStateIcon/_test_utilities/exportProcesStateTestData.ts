import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllExportProcessStatePermutations = (): ModelInfoTypes.ExportProcessState[] => {
  // generate all permutations of ExportProcessState status, errored, exportErrors, exportWarnings
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ExportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((exportErrors) => {
        allBooleans.forEach((exportWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              exportErrors: exportErrors,
              exportWarnings: exportWarnings,
            },
            downloadUrl: "https://example.com",
            timestamp: new Date(),
          });
        });
      });
    });
  });
  return allPermutations;
};
