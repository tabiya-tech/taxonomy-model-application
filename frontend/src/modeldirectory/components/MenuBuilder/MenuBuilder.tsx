import React, { useContext } from "react";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ModelInfo = ModelInfoTypes.ModelInfo;
import { Icon, ListItemIcon, ListItemText, Menu, MenuItem, Typography } from "@mui/material";
import { IsOnlineContext } from "src/app/providers";

const uniqueId = "2d06ac7f-1dd6-456a-97d8-54f34858d727";

export const DATA_TEST_ID = {
  MENU: `${uniqueId}-menu`,
  MENU_ITEM: `${uniqueId}-menu-item`,
  MENU_ITEM_ICON: `${uniqueId}-menu-item-icon`,
  MENU_ITEM_TEXT: `${uniqueId}-menu-item-text`,
};

export interface MenuItemConfig {
  text: string;
  icon: React.ReactNode;
  disableOnNonSuccessfulImport?: boolean;
  disableWhenOffline?: boolean;
  onClick: (modelId: string) => void;
}

export interface MenuBuilderProps {
  items: MenuItemConfig[];
  open: boolean;
  anchorEl: HTMLElement | null;
  model: ModelInfo | null;
  notifyOnClose: () => void;
}

export const isModelNotSuccessfullyImported = (model: ModelInfo | null) => {
  return (
    model?.importProcessState?.status !== ImportProcessStateAPISpecs.Enums.Status.COMPLETED ||
    model?.importProcessState?.result.errored !== false
  );
};

function MenuBuilder(props: Readonly<MenuBuilderProps>) {
  const isOnline = useContext(IsOnlineContext);

  const isItemDisabled = (item: MenuItemConfig) => {
    if (!isOnline && item.disableWhenOffline) return true;
    return item.disableOnNonSuccessfulImport && isModelNotSuccessfullyImported(props.model);
  };

  const handleItemClick = (item: MenuItemConfig) => {
    props.notifyOnClose();
    item.onClick(props.model!.id);
  };

  return (
    <Menu
      data-testid={DATA_TEST_ID.MENU}
      elevation={2}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
    >
      {props.items.map((item, index) => (
        <MenuItem
          onClick={() => handleItemClick(item)}
          component="button"
          data-testid={DATA_TEST_ID.MENU_ITEM}
          disabled={isItemDisabled(item)}
          key={index}
        >
          <ListItemIcon data-testid={DATA_TEST_ID.MENU_ITEM_ICON}>
            <Icon color="primary">{item.icon}</Icon>
          </ListItemIcon>
          <ListItemText data-testid={DATA_TEST_ID.MENU_ITEM_TEXT}>
            <Typography variant={"caption"} color={"primary"}>
              {item.text}
            </Typography>
          </ListItemText>
        </MenuItem>
      ))}
    </Menu>
  );
}

export default MenuBuilder;
