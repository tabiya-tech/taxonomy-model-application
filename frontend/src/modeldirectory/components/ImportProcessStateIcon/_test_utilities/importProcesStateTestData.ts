import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllImportProcessStatePermutations = (): ModelInfoTypes.ImportProcessState[] => {
  // generate all permutations of ImportProcessState status, errored, parsingErrors, parsingWarnings
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ImportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((parsingErrors) => {
        allBooleans.forEach((parsingWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              parsingErrors: parsingErrors,
              parsingWarnings: parsingWarnings,
            },
            createdAt: new Date(allPermutations.length),
            updatedAt: new Date(allPermutations.length),
          });
        });
      });
    });
  });
  return allPermutations;
};
