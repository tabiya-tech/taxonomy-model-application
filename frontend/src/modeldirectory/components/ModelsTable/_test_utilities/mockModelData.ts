import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { ImportProcessStateEnums } from "api-specifications/importProcessState/enums";

import { v4 as uuidv4 } from "uuid";

import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { getMockId } from "src/_test_utilities/mockMongoId";
import { getRandomLorem, getRandomString, getTestString } from "src/_test_utilities/specialCharacters";
import { faker } from "@faker-js/faker";
import { CELL_MAX_LENGTH } from "src/modeldirectory/components/ModelsTable/ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  const model = getArrayOfFakeModels(1)[0];
  model.id = getMockId(id ?? 1);
  return model;
}

export function getArrayOfFakeModels(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location
          .countryCode("alpha-3")
          .substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description:
        i % 2 === 0
          ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH)
          : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getArrayOfFakeModelsMaxLength(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export const fakeModel: ModelInfoTypes.ModelInfo = {
  UUID: "8d914eab-6f7d-4183-accc-a09f99887b39",
  UUIDHistory: ["8d914eab-6f7d-4183-accc-a09f99887b39"],
  description: "aw j zt   agrkasl dy ogtimpsauwumu l utrovthao syertm beawpxluhyudgzbbm",
  id: "000000000000000000000001",
  importProcessState: {
    id: "000000000000000000000001",
    result: {
      errored: true,
      parsingErrors: true,
      parsingWarnings: true,
    },
    status: ImportProcessStateEnums.Status.PENDING,
    createdAt: new Date("2023-10-18T17:35:10.571Z"),
    updatedAt: new Date("2023-10-18T17:35:12.571Z"),
  },
  exportProcessState: [
    {
      downloadUrl: "https://www.example.com/",
      id: "000000000000000000000001",
      result: {
        errored: true,
        exportErrors: true,
        exportWarnings: true,
      },
      status: ExportProcessStateAPISpecs.Enums.Status.PENDING,
      timestamp: new Date("2023-10-18T17:35:10.571Z"),
      createdAt: new Date("2023-10-18T17:35:10.571Z"),
      updatedAt: new Date("2023-10-18T17:35:12.571Z"),
    },
  ],
  locale: {
    UUID: "7c4ea00b-fbea-4a31-9c04-009c813f5f34",
    name: "Palestine",
    shortCode: "ETH",
  },
  name: "1/1 - fysmros tcgcnjbbrrev",
  path: "https://unwelcome-editorial.net/",
  releaseNotes: "Perferendis modi impedit necessitatibus a",
  released: true,
  tabiyaPath: "https://silver-vulture.biz",
  createdAt: new Date("2023-10-18T17:35:10.571Z"),
  updatedAt: new Date("2022-10-18T17:35:10.571Z"),
  version: "0.2.2",
};

export function getRandomImportStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getRandomExportStatus(id: number) {
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getOneFakeExportProcessState(i: number): ModelInfoTypes.ExportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomExportStatus(i),
    result: {
      errored: i % 2 === 0,
      exportErrors: i % 2 === 0,
      exportWarnings: i % 2 === 0,
    },
    downloadUrl: faker.internet.url(),
    timestamp: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}

export function getOneFakeImportProcessState(i: number): ModelInfoTypes.ImportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomImportStatus(i),
    result: {
      errored: i % 2 === 0,
      parsingErrors: i % 2 === 0,
      parsingWarnings: i % 2 === 0,
    },
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}
