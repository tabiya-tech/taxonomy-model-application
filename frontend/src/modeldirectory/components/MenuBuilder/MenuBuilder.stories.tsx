import { Meta, StoryObj } from "@storybook/react";
import MenuBuilder, { MenuBuilderProps } from "./MenuBuilder";
import * as React from "react";
import { useEffect } from "react";
import Box from "@mui/material/Box";
import { action } from "@storybook/addon-actions";
import CloudDownloadIcon from "@mui/icons-material/CloudDownload";
import { SendOutlined } from "@mui/icons-material";
import { getOneRandomModelMaxLength } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";

const meta: Meta<typeof MenuBuilder> = {
  title: "ModelDirectory/MenuBuilder",
  component: MenuBuilder,
  tags: ["autodocs"],
  argTypes: {
    open: {
      control: {
        type: "boolean",
      },
    },
    notifyOnClose: {
      table: {
        disable: true,
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof MenuBuilder>;

export const Shown: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    open: true,
    items: [
      {
        text: "Item 1",
        icon: <CloudDownloadIcon />,
        onClick: action("Item 1 clicked"),
      },
      {
        text: "Item 2",
        icon: <SendOutlined />,
        onClick: action("Item 2 clicked"),
      },
    ],
  },
};

export const DisabledOptions: Story = {
  render: (args) => {
    const model = getOneRandomModelMaxLength();
    model.importProcessState.status = ImportProcessStateAPISpecs.Enums.Status.PENDING;
    return <SetupComponent {...args} model={model} />;
  },
  args: {
    open: true,
    items: [
      {
        text: "Item 1",
        icon: <CloudDownloadIcon />,
        onClick: action("Item 1 clicked"),
        disableOnNonSuccessfulImport: true,
      },
      {
        text: "Item 2",
        icon: <SendOutlined />,
        onClick: action("Item 2 clicked"),
        disableOnNonSuccessfulImport: true,
      },
    ],
  },
};

function SetupComponent(props: Readonly<MenuBuilderProps>) {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  useEffect(() => {
    const anchor = document.getElementById("anchor-on-me");
    setAnchorEl(anchor);
  }, []);

  return (
    <>
      <Box display="flex" justifyContent="center" alignItems="center" height="100%">
        <div style={{ border: "1px dashed", height: "20px", width: "100px" }} id="anchor-on-me" />
      </Box>
      <MenuBuilder
        anchorEl={anchorEl}
        open={anchorEl !== null && props.open}
        notifyOnClose={action("notifyOnClose")}
        items={props.items}
        model={props.model}
      />
    </>
  );
}
