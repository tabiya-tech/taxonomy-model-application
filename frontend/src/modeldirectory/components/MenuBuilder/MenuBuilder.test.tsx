// mute the console
import "src/_test_utilities/consoleMock";

import MenuBuilder, { DATA_TEST_ID } from "./MenuBuilder";
import { fireEvent, screen, waitFor, within } from "@testing-library/react";
import { render } from "src/_test_utilities/test-utils";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getOneRandomModelMaxLength } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import { mockBrowserIsOnLine, unmockBrowserIsOnLine } from "src/_test_utilities/mockBrowserIsOnline";

describe("MenuBuilder", () => {
  const stdGivenProps = {
    open: true,
    anchorEl: document.createElement("div"), // Mock an HTMLElement
    model: null,
    notifyOnClose: jest.fn(),
  };

  describe("Render tests", () => {
    beforeEach(() => {
      unmockBrowserIsOnLine();
      (console.error as jest.Mock).mockClear();
      (console.warn as jest.Mock).mockClear();
    });

    test("render the menu visible", () => {
      // GIVEN some menu items
      const givenItems = [
        {
          text: "Item 1",
          icon: <div>icon1</div>,
          onClick: jest.fn(),
        },
        {
          text: "Item 2",
          icon: <div>icon2</div>,
          onClick: jest.fn(),
        },
      ];

      // WHEN the MenuBuilder is rendered
      render(<MenuBuilder {...stdGivenProps} items={givenItems} />);

      // THEN expect it to be shown
      const menuElement = screen.getByTestId(DATA_TEST_ID.MENU);
      expect(menuElement).toBeInTheDocument();

      // AND to have the correct number of items
      const menuItems = screen.getAllByTestId(DATA_TEST_ID.MENU_ITEM);
      expect(menuItems).toHaveLength(givenItems.length);
      // AND every item to have the correct text and icon
      menuItems.forEach((item, index) => {
        const itemTextElement = within(item).getByTestId(DATA_TEST_ID.MENU_ITEM_TEXT);
        expect(itemTextElement).toBeInTheDocument();
        expect(itemTextElement).toHaveTextContent(givenItems[index].text);
        // AND the icon to be present
        const itemIconElement = within(item).getByTestId(DATA_TEST_ID.MENU_ITEM_ICON);
        expect(itemIconElement).toBeInTheDocument();
      });
      // AND to match the snapshot
      expect(menuElement).toMatchSnapshot();
      // AND no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });

    test("renders the menu hidden", () => {
      // GIVEN some menu items
      const givenItems = [
        {
          text: "Item 1",
          icon: <div>icon1</div>,
          onClick: jest.fn(),
        },
        {
          text: "Item 2",
          icon: <div>icon2</div>,
          onClick: jest.fn(),
        },
      ];

      // WHEN the MenuBuilder is rendered
      render(<MenuBuilder {...stdGivenProps} items={givenItems} open={false} />);

      // THEN expect it to be hidden
      const menuElement = screen.queryByTestId(DATA_TEST_ID.MENU);
      expect(menuElement).not.toBeInTheDocument();
      // AND no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });

    describe("Browser offline/online tests", () => {
      test.each([
        [false, { disableWhenOffline: true, isOnline: true }],
        [true, { disableWhenOffline: true, isOnline: false }],
        [false, { disableWhenOffline: false, isOnline: true }],
        [false, { disableWhenOffline: false, isOnline: false }],
      ])(`renders the menu item disabled = %s when %j`, (expectedState, testCase) => {
        // GIVEN some menu items
        const givenItems = [
          {
            text: "Item 1",
            icon: <div>icon1</div>,
            onClick: jest.fn(),
            disableWhenOffline: testCase.disableWhenOffline,
          },
        ];
        // AND the browser is online/offline
        mockBrowserIsOnLine(testCase.isOnline);

        // WHEN the MenuBuilder is rendered
        render(<MenuBuilder {...stdGivenProps} items={givenItems} />);

        // THEN expect the menu item to be disabled
        const menuItem = screen.getByTestId(DATA_TEST_ID.MENU_ITEM);
        expect(menuItem).toBeInTheDocument();
        const matcher = expectedState ? "toBeDisabled" : "toBeEnabled";
        expect(menuItem)[matcher]();
        // AND no errors or warning to have occurred
        expect(console.error).not.toHaveBeenCalled();
        expect(console.warn).not.toHaveBeenCalled();
      });

      test("renders the menu item enabled->disabled->enabled when online status changes and disableWhenOffline is true", async () => {
        // GIVEN that the internet status is online
        mockBrowserIsOnLine(true);
        // AND a menu item
        const givenItems = [
          {
            text: "Item 1",
            icon: <div>icon1</div>,
            onClick: jest.fn(),
            disableWhenOffline: true,
          },
        ];

        // WHEN the MenuBuilder is rendered
        render(<MenuBuilder {...stdGivenProps} items={givenItems} />);

        // THEN expect the menu item to be enabled
        const menuItem = screen.getByTestId(DATA_TEST_ID.MENU_ITEM);
        expect(menuItem).toBeEnabled();

        // WHEN the internet status changes to offline
        mockBrowserIsOnLine(false);

        // THEN expect the menu item to be disabled
        expect(menuItem).toBeDisabled();

        // WHEN the internet status changes back to online
        mockBrowserIsOnLine(true);

        // THEN expect the menu item to be enabled
        expect(menuItem).toBeEnabled();
      });
    });

    describe("Import process state successful/non-successful tests", () => {
      test.each([
        [true, ImportProcessStateAPISpecs.Enums.Status.PENDING, null],
        [true, ImportProcessStateAPISpecs.Enums.Status.RUNNING, null],
        [
          true,
          ImportProcessStateAPISpecs.Enums.Status.COMPLETED,
          {
            parsingErrors: false,
            parsingWarnings: false,
            errored: true,
          },
        ],
        [
          false,
          ImportProcessStateAPISpecs.Enums.Status.COMPLETED,
          {
            parsingErrors: true,
            parsingWarnings: false,
            errored: false,
          },
        ],
        [
          false,
          ImportProcessStateAPISpecs.Enums.Status.COMPLETED,
          {
            parsingErrors: false,
            parsingWarnings: true,
            errored: false,
          },
        ],
        [
          false,
          ImportProcessStateAPISpecs.Enums.Status.COMPLETED,
          {
            parsingErrors: true,
            parsingWarnings: true,
            errored: false,
          },
        ],
      ])(
        "renders the menu item disabled =  %s when the model import status is %d and the result %j",
        (expectedState, givenStatus, givenResult) => {
          // GIVEN some menu items
          const givenItems = [
            {
              text: "Item 1",
              icon: <div>icon1</div>,
              onClick: jest.fn(),
              disableOnNonSuccessfulImport: true,
            },
          ];
          // AND a model with the given status and result
          const givenModel = getOneRandomModelMaxLength();
          givenModel.importProcessState.status = givenStatus;
          if (givenResult) {
            givenModel.importProcessState.result = givenResult;
          }

          // WHEN the MenuBuilder is rendered
          render(<MenuBuilder {...stdGivenProps} items={givenItems} model={givenModel} />);

          // THEN expect the menu item to be disabled
          const menuItem = screen.getByTestId(DATA_TEST_ID.MENU_ITEM);
          expect(menuItem).toBeInTheDocument();
          const matcher = expectedState ? "toBeDisabled" : "toBeEnabled";
          expect(menuItem)[matcher]();
          // AND no errors or warning to have occurred
          expect(console.error).not.toHaveBeenCalled();
          expect(console.warn).not.toHaveBeenCalled();
        }
      );
    });

    describe("Browser offline/online and Import process state successful/non-successful tests", () => {
      const successfulResult = { errored: false, parsingErrors: false, parsingWarnings: false };
      const NonSuccessfulResult = { errored: true, parsingErrors: false, parsingWarnings: false };
      enum BROWSER_STATE {
        ONLINE = "online",
        OFFLINE = "offline",
      }
      enum IMPORT_STATE {
        SUCCESSFUL = "successful",
        NON_SUCCESSFUL = "nonSuccessful",
      }
      test.each([
        [true, BROWSER_STATE.OFFLINE, IMPORT_STATE.SUCCESSFUL],
        [true, BROWSER_STATE.OFFLINE, IMPORT_STATE.NON_SUCCESSFUL],
        [true, BROWSER_STATE.ONLINE, IMPORT_STATE.NON_SUCCESSFUL],
        [false, BROWSER_STATE.ONLINE, IMPORT_STATE.SUCCESSFUL],
      ])(
        `renders the menu item disabled =  %s when the browser is %s and the model import status is %s`,
        (expectedState, givenBrowserState, givenImportState) => {
          // GIVEN some menu items
          const givenItems = [
            {
              text: "Item 1",
              icon: <div>icon1</div>,
              onClick: jest.fn(),
              disableWhenOffline: true,
              disableOnNonSuccessfulImport: true,
            },
          ];
          // AND the browser is online/offline
          mockBrowserIsOnLine(givenBrowserState === BROWSER_STATE.ONLINE);
          // AND a model with the given status and result
          const givenModel = getOneRandomModelMaxLength();
          givenModel.importProcessState.status = ImportProcessStateAPISpecs.Enums.Status.COMPLETED;
          givenModel.importProcessState.result =
            givenImportState === IMPORT_STATE.SUCCESSFUL ? successfulResult : NonSuccessfulResult;

          // WHEN the MenuBuilder is rendered
          render(<MenuBuilder {...stdGivenProps} items={givenItems} model={givenModel} />);

          // THEN expect the menu item to be disabled
          const menuItem = screen.getByTestId(DATA_TEST_ID.MENU_ITEM);
          expect(menuItem).toBeInTheDocument();
          const matcher = expectedState ? "toBeDisabled" : "toBeEnabled";
          expect(menuItem)[matcher]();
          // AND no errors or warning to have occurred
          expect(console.error).not.toHaveBeenCalled();
          expect(console.warn).not.toHaveBeenCalled();
        }
      );
    });
  });

  describe("Action tests", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test("should call notifyOnClose then the menu item's onClick function with the modelId when the menu item is clicked", async () => {
      // GIVEN an onClick function for the menu item
      const givenOnClickFn = jest.fn();
      // AND some menu items
      const givenItems = [
        {
          text: "Item 1",
          icon: <div>icon1</div>,
          onClick: givenOnClickFn,
        },
      ];
      // AND a current model
      const givenModel = getOneRandomModelMaxLength();
      // AND the MenuBuilder is rendered
      render(<MenuBuilder {...stdGivenProps} items={givenItems} model={givenModel} />);

      // WHEN the menu item is clicked
      fireEvent.click(screen.getByTestId(DATA_TEST_ID.MENU_ITEM));

      // THEN expect the onClick function should be called
      expect(givenOnClickFn).toHaveBeenCalledWith(givenModel.id);

      // THEN the notifyOnClose should be called to close the menu
      await waitFor(() => expect(stdGivenProps.notifyOnClose).toHaveBeenCalled());
      // AND the onClick should be called after the menu is closed
      expect(givenOnClickFn).toHaveBeenCalledAfter(stdGivenProps.notifyOnClose);
    });

    test("should successfully close the menu when esc is pressed", async () => {
      // GIVEN that the menu is rendered
      render(<MenuBuilder {...stdGivenProps} items={[]} />);

      // WHEN the esc key is pressed
      const menuElement = screen.getByTestId(DATA_TEST_ID.MENU);
      fireEvent.keyDown(menuElement, { key: "Escape" });

      // THEN the notifyOnClose should be called to close the menu
      await waitFor(() => expect(stdGivenProps.notifyOnClose).toHaveBeenCalled());
    });
  });
});
