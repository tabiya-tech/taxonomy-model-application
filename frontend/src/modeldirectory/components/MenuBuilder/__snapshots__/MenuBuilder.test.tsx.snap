// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuBuilder Render tests render the menu visible 1`] = `
<div
  class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
  data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiMenu-paper MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiPopover-paper css-76nwsu-MuiPaper-root-MuiMenu-paper-MuiPopover-paper"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
      role="menu"
      tabindex="-1"
    >
      <button
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-ram008-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item-icon"
        >
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-colorPrimary MuiIcon-fontSizeMedium css-140lp5m-MuiIcon-root"
          >
            <div>
              icon1
            </div>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item-text"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1w1i5dg-MuiTypography-root"
          >
            Item 1
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-ram008-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item-icon"
        >
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-colorPrimary MuiIcon-fontSizeMedium css-140lp5m-MuiIcon-root"
          >
            <div>
              icon2
            </div>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          data-testid="2d06ac7f-1dd6-456a-97d8-54f34858d727-menu-item-text"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1w1i5dg-MuiTypography-root"
          >
            Item 2
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </ul>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
